File: glean/glass/Glean/Glass/Path.hs
{-
  Copyright (c) Meta Platforms, Inc. and affiliates.
  All rights reserved.

  This source code is licensed under the BSD-style license found in the
  LICENSE file in the root directory of this source tree.
-}

module Glean.Glass.Path where

import qualified Data.Text as Text

import Glean.Glass.Base ( SymbolRepoPath(..), GleanPath(GleanPath) )
import qualified Glean.Glass.Types as Glass

-- TODO: We probably want to read the repo config from a file, or put it in a
-- server config

-- Convert repo-relative Glass normalized paths to Glean-index specific paths
toGleanPath :: SymbolRepoPath -> GleanPath
toGleanPath (SymbolRepoPath _ (Glass.Path path)) = GleanPath path

-- | Site-level rules for processing index paths to the filesystem
-- Glass paths are always repo-root relative.
fromGleanPath
  :: Glass.RepoName -> GleanPath -> SymbolRepoPath
fromGleanPath repo (GleanPath path) = SymbolRepoPath repo $ Glass.Path path

File: glean/lang/kotlin/indexer/glean_utils/Path.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

package glean.lang.kotlin.indexer.glean_utils

import com.facebook.glean.schema.javakotlin_alpha.Path
import com.facebook.glean.schema.javakotlin_alpha.PathKey

fun buildPath(pathSegments: List<org.jetbrains.kotlin.name.Name>): Path? {
  if (pathSegments.isEmpty()) {
    return null
  }
  var keyBuilder = PathKey.Builder().setBase(buildName(pathSegments.last().identifier))

  if (pathSegments.size > 1) {
    val subPath = buildPath(pathSegments.subList(0, pathSegments.size - 2))
    keyBuilder = keyBuilder.setContainer(subPath)
  }
  return Path.Builder().setKey(keyBuilder.build()).build()
}
